dadb/src/main/kotlin/dadb/forwarding/TcpForwarder.kt

```diff
@@ -0,0 +1,133 @@
+package dadb.forwarding
+
+import dadb.Dadb
+import okio.BufferedSink
+import okio.Source
+import okio.buffer
+import okio.sink
+import okio.source
+import java.io.InterruptedIOException
+import java.net.ServerSocket
+import kotlin.concurrent.thread
+
+internal class TcpForwarder(
+    private val dadb: Dadb,
+    private val hostPort: Int,
+    private val targetPort: Int,
+) : AutoCloseable {
+
+    private var state: State = State.STOPPED
+    private var thread: Thread? = null
+
+    fun start() {
+        if (state != State.STOPPED) {
+            throw IllegalStateException("Forwarder is already started at port $hostPort")
+        }
```

Comment:
You can use `check()` to make it more idiomatic if you like as this is a typical example of its usage:

```kotlin
check(state == State.STOPPED) { "Forwarder is already started at port $hostPort" }
```